File: 20230713_Thursday/20230713.rb
# Leetcode: 207. Course Schedule.
# https://leetcode.com/problems/course-schedule/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 104 ms, faster than 35.18% of Ruby online submissions for Course
# Schedule.
# Memory Usage: 212.7 MB, less than 31.48% of Ruby online submissions for Course
# Schedule.
# 2023.07.13 Daily Challenge.
# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Boolean}
def can_finish(num_courses, prerequisites)
  @ps = prerequisites
  # Based on:
  # https://leetcode.com/problems/course-schedule/solution/
  @adj = Array.new(num_courses){ Array.new()}
  @ps.each do |pt|
    @adj[pt[1]].push(pt[0])
  end
  @v = Array.new(num_courses,false)
  @is = Array.new(num_courses,false)
  (0...num_courses).each do |i|
    return false if dfs(i)
  end
  return true
end

def dfs(i)
  return true if @is[i]
  return false if @v[i]
  @v[i] = true
  @is[i] = true
  @adj[i].each do |j|
    return true if dfs(j)
  end
  @is[i] = false
  return false
end

File: 20230713_Thursday/20230713.go
// Leetcode: 207. Course Schedule.
// https://leetcode.com/problems/course-schedule/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 11 ms, faster than 75.12% of Go online submissions for Course
// Schedule.
// Memory Usage: 6.2 MB, less than 75.62% of Go online submissions for Course
// Schedule.
// 2023.07.14 Updated.
package main

import (
)

var n int
var prs [][]int
var adj [][]int
var v []bool
var is []bool

func canFinish(numCourses int, prerequisites [][]int) bool {
	prs = prerequisites
	n = numCourses 
	adj = make([][]int, n)
	v = make([]bool, n)
	is = make([]bool, n)
	for _, pr := range prs {
		a, b := pr[1], pr[0]
		if nil == adj[a] {
			adj[a] = make([]int, 0)
		}
		adj[a] = append(adj[a], b)
	}
	for i := 0; i < n; i += 1 {
		if dfs(i) {
			return false
		}
	}
	return true
}

func dfs(i int) bool {
	if is[i] {
		return true
	}
	if v[i] {
		return false
	}
	v[i] = true
	is[i] = true
	for _, j := range adj[i] {
		if dfs(j) {
			return true
		}
	}
	is[i] = false
	return false
}
